From 39a3cd0f4210dbec1ae8ef39a87d34b76eac91a3 Mon Sep 17 00:00:00 2001
From: Nezih Yigitbasi <nyigitbasi@netflix.com>
Date: Mon, 25 Apr 2016 15:05:11 -0700
Subject: [PATCH] PARQUET-560: Synchronize writes to the finishCalled variable

Reads of the `finishCalled` variable are properly synchronized, but writes are not -- so there's some sort of inconsistent synch. going on here. This PR fixes that.

/cc @rdblue can you please take a look?

Author: Nezih Yigitbasi <nyigitbasi@netflix.com>

Closes #334 from nezihyigitbasi/sc-synch-fix and squashes the following commits:

a85cf0c [Nezih Yigitbasi] Synchronize writes to the finishCalled variable
---
 .../java/org/apache/parquet/hadoop/codec/SnappyCompressor.java  | 2 +-
 pom.xml                                                         | 1 +
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/codec/SnappyCompressor.java b/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/codec/SnappyCompressor.java
index f09989634a..d0270ca7c1 100644
--- a/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/codec/SnappyCompressor.java
+++ b/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/codec/SnappyCompressor.java
@@ -113,7 +113,7 @@ public void end() {
   }
 
   @Override
-  public void finish() {
+  public synchronized void finish() {
     finishCalled = true;
   }
 
diff --git a/pom.xml b/pom.xml
index 9314a532e8..512bf3735e 100644
--- a/pom.xml
+++ b/pom.xml
@@ -245,6 +245,7 @@
                      <exclude>org/apache/parquet/avro/SpecificDataSupplier</exclude> <!-- made public -->
                      <exclude>org/apache/parquet/io/ColumnIOFactory$ColumnIOCreatorVisitor</exclude> <!-- removed non-API class -->
                      <exclude>org/apache/parquet/io/ColumnIOFactory/**</exclude> <!-- removed non-API class and methods-->
+		     <exclude>org/apache/parquet/hadoop/codec/SnappyCompressor</exclude> <!-- added synchronized modifier -->
                    </excludes>
                  </requireBackwardCompatibility>
                </rules>
